### ISSUE TITLE
`batchtools` build: HTTP 404 NOT FOUND

### ISSUE THREAD

**Issue by nick-youngblut** [conda/conda-build#2807]
I'm getting the following error when trying to build batchtools:

```
## Package Plan ##

  environment location: /ebio/abt3_projects/software/dev/miniconda3_dev/conda-bld/r-base64url_1523010509371/_build_env


The following NEW packages will be INSTALLED:

    binutils_impl_linux-64: 2.28.1-had2808c_3
    binutils_linux-64:      7.2.0-26
    gcc_impl_linux-64:      7.2.0-habb00fd_3
    gcc_linux-64:           7.2.0-26
    libstdcxx-ng:           7.2.0-hdf63c60_3
    make:                   4.2.1-0           conda-forge

Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
Source cache directory is: /ebio/abt3_projects/software/dev/miniconda3_dev/conda-bld/src_cache
Downloading source to cache: base64url_1.3_0db22ef6d3.tar.gz
Downloading https://mran.microsoft.com/snapshot/2018-01-01/src/contrib/base64url_1.3.tar.gz
Error: HTTP 404 NOT FOUND for url <https://mran.microsoft.com/snapshot/2018-01-01/src/contrib/base64url_1.3.tar.gz>
Elapsed: 00:00.727542

An HTTP error occurred when trying to retrieve this URL.
HTTP errors are often intermittent, and a simple retry will get you on your way.
Downloading https://mran.microsoft.com/snapshot/2018-01-01/src/contrib/Archive/base64url/base64url_1.3.tar.gz
Error: HTTP 404 NOT FOUND for url <https://mran.microsoft.com/snapshot/2018-01-01/src/contrib/Archive/base64url/base64url_1.3.tar.gz>
Elapsed: 00:00.704098

An HTTP error occurred when trying to retrieve this URL.
HTTP errors are often intermittent, and a simple retry will get you on your way.
Could not download https://mran.microsoft.com/snapshot/2018-01-01/src/contrib/Archive/base64url/base64url_1.3.tar.gz
```

If I removed "bioconda" and "conda-forge" from my `~/.condarc`, then I get the following error:

```
$ conda-build r-batchtools/
Adding in variants from internal_defaults
INFO:conda_build.variants:Adding in variants from internal_defaults
Attempting to finalize metadata for r-batchtools
INFO:conda_build.metadata:Attempting to finalize metadata for r-batchtools
Solving environment: ...working... done
Solving environment: ...working... failed
Adding in variants from config.variant
INFO:conda_build.variants:Adding in variants from config.variant
Missing dependency r-base64url, but found recipe directory, so building r-base64url first
Missing dependency r-progress, but found recipe directory, so building r-progress first
Attempting to finalize metadata for r-progress
INFO:conda_build.metadata:Attempting to finalize metadata for r-progress
Solving environment: ...working... failed
failed to get install actions, retrying.  exception was: The following specifications were found to be in conflict:
  - r-base=3.4
  - r-prettyunits -> r=3.3.1 -> r-base==3.3.1
Use "conda info <package>" to see the dependencies for each package.
WARNING:conda_build.environ:failed to get install actions, retrying.  exception was: The following specifications were found to be in conflict:
  - r-base=3.4
  - r-prettyunits -> r=3.3.1 -> r-base==3.3.1
Use "conda info <package>" to see the dependencies for each package.
```

Conda info:

```
$ conda info

     active environment : R_pkg_build
    active env location : /ebio/abt3_projects/software/dev/miniconda3_dev/envs/R_pkg_build
            shell level : 1
       user config file : /ebio/abt3/nyoungblut/.condarc
 populated config files : /ebio/abt3_projects/software/dev/miniconda3_dev/.condarc
                          /ebio/abt3/nyoungblut/.condarc
          conda version : 4.5.0
    conda-build version : 3.8.0
         python version : 3.6.5.final.0
       base environment : /ebio/abt3_projects/software/dev/miniconda3_dev  (writable)
           channel URLs : https://conda.anaconda.org/bioconda/linux-64
                          https://conda.anaconda.org/bioconda/noarch
                          https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/free/linux-64
                          https://repo.anaconda.com/pkgs/free/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/pro/linux-64
                          https://repo.anaconda.com/pkgs/pro/noarch
                          https://conda.anaconda.org/r/linux-64
                          https://conda.anaconda.org/r/noarch
                          https://conda.anaconda.org/qiime2/linux-64
                          https://conda.anaconda.org/qiime2/noarch
          package cache : /ebio/abt3_projects/software/dev/miniconda3_dev/pkgs
                          /ebio/abt3/nyoungblut/.conda/pkgs
       envs directories : /ebio/abt3_projects/software/dev/miniconda3_dev/envs
                          /ebio/abt3/nyoungblut/.conda/envs
               platform : linux-64
             user-agent : conda/4.5.0 requests/2.18.4 CPython/3.6.5 Linux/4.4.67 ubuntu/16.04 glibc/2.23
                UID:GID : 6354:350
             netrc file : None
           offline mode : False
```

**Comment by msarahan** [comment]
The skeleton got a bad URL somehow.  Either MRAN isn't mirroring everything, or the MRAN mirror may be out of date somehow?

CC @mingwandroid

**Comment by nick-youngblut** [comment]
I just got a similar message when trying to build `boruta`:

```
Source cache directory is: /ebio/abt3_projects/software/dev/miniconda3_dev/conda-bld/src_cache
Downloading source to cache: Boruta_5.3.0_29366cc13c.tar.gz
Downloading https://mran.microsoft.com/snapshot/2018-01-01/src/contrib/Boruta_5.3.0.tar.gz
Error: HTTP 404 NOT FOUND for url <https://mran.microsoft.com/snapshot/2018-01-01/src/contrib/Boruta_5.3.0.tar.gz>
Elapsed: 00:00.764033

An HTTP error occurred when trying to retrieve this URL.
HTTP errors are often intermittent, and a simple retry will get you on your way.
Downloading https://mran.microsoft.com/snapshot/2018-01-01/src/contrib/Archive/Boruta/Boruta_5.3.0.tar.gz
Error: HTTP 404 NOT FOUND for url <https://mran.microsoft.com/snapshot/2018-01-01/src/contrib/Archive/Boruta/Boruta_5.3.0.tar.gz>
Elapsed: 00:01.023688

An HTTP error occurred when trying to retrieve this URL.
HTTP errors are often intermittent, and a simple retry will get you on your way.
Could not download https://mran.microsoft.com/snapshot/2018-01-01/src/contrib/Archive/Boruta/Boruta_5.3.0.tar.gz
```

Should I revert to a previous version of `conda-build`?

**Comment by mingwandroid** [comment]
These are not bugs. You have generated recipes against CRAN which has more recent releases on it and are then trying to build them against MRAN.

**Comment by mingwandroid** [comment]
`cran boruta --cran-url https://mran.microsoft.com/snapshot/2018-01-01`

**Comment by mingwandroid** [comment]
@nick-youngblut, please try having a read at the output of `conda skeleton cran --help` sometime.

**Comment by mingwandroid** [comment]
If you want to build against CRAN then add:

```
cran_mirror:
 - https://cran.r-project.org
```

to conda_build_config_yaml and pass that to `conda-build` as `conda-build -m conda_build_config_yaml`

**Comment by nick-youngblut** [comment]
> @nick-youngblut, please try having a read at the output of conda skeleton cran --help sometime

I have looked at the documentation many times, but I don't see how it explicitly states why this CRAN/MRAN issue occurred, and the error:

```
An HTTP error occurred when trying to retrieve this URL.
HTTP errors are often intermittent, and a simple retry will get you on your way
```

Is misleading.

Sure, I could spend more time studying the documentation and looking into the build errors more, but I'm also trying to get research done. Many of my colleagues at my institution have also be trying to use Jupyter + R, but most have found it too much of hassle and have just switched to using Rstudio for all work in R. Most have completely abandoned using Jupyter + conda.  I've been holding out on that, but I think that I'm going to do the same, so you probably won't get any more `conda-build` issue reports from me, at least not for R packages.

**Comment by mingwandroid** [comment]
OK, thanks for letting me know.

**Comment by mingwandroid** [comment]
BTW we provide RStudio too, the issues you have with R stem from the fact that basically no one builds binary R package for Linux (except Debian and Fedora AFAIK, but AFAIK even those are not installed via install.packages() from the REPL).

So you are left in a situation where you need to compile some of them yourself and that is complicated and not really something that data scientists should have to deal with.

That's why the Anaconda Distribution builds binaries for R in the first place (and remember we *are* a binary distribution), but we can only build so many of them. The conda-build `skeleton` commands are primarily tools for us and conda-forge to use to build packages and as such are developed in a very ad-hoc way. For example I will not touch `conda skeleton cran` until I go to update the entire infrastructure of R packages at each release of a new R version upstream and as soon as I build those packages I move on to other more pressing issues.

If you want to help out to make these tools more refined then that would be appreciated, but from past experience we will have to bend them to our specific needs very quickly, and often to the detriment of their quality. As an example I added MRO support recently but had no time to refactor it as I would have liked to. I think that all of `conda skeleton` needs to be rewritten so that a core of it is shared between all the language ecosystems, but as I say our focus here is on building binary packages.

**Comment by msarahan** [comment]
@nick-youngblut I hope you won't give up on this, but I certainly understand your frustration.  We might be able to address this issue by adding two source urls to the cran skeleton: first the MRAN one, then CRAN.

I really hope you won't give up on this because the main reason why our R stuff is so rough is that none of us are R experts.  Having people like you help us understand the rough spots that we need to improve is very helpful.  That said, I know how frustrating it can be to always feel like a guinea pig.

**Comment by nick-youngblut** [comment]
> BTW we provide RStudio too

While it is great that you provide RStudio, my research group and the other research groups at my institute really need RStudio-Server to effectively utilize our computational infrastructure.

> If you want to help out to make these tools more refined then that would be appreciated

I would love to help out, but my lack of knowledge in this area would probably lead to more problems than solutions.

> I hope you won't give up on this

Myself and other members of my research group really like Jupyter, conda, and the ability to use R with both, but just out of practicality, we need to spend our time analyzing data instead of installing software (which conda is supposed to help with). I keep having problems with `conda-build` (eg., I've been trying to get the `fd` package built for weeks). Many members of my research group are biologists with limited computer science knowledge, so they really need something that "just works". 

> none of us are R experts

It's really too bad that there's such a big divide between R and python developers. It can get a bit frustrating for researchers who just want to use whatever tool is best for their particular research question, but often these tools don't work well with each other.

**Comment by mingwandroid** [comment]
>  I've been trying to get the fd package built for weeks

> we need to spend our time analyzing data instead of installing software (which conda is supposed to help with)

.. we do that by providing binary packages, conda is after all, an installer of binary packages primarily. Anything beyond that is a best-effort as time allows.

Perhaps conda-forge and/or bioconda could be better suited to your particular needs here? They provide more prebuilt R packages than we do at present and have some experts in package building. The stuff in AD is more modern and faster but there's less of it.

If you want us to help out with your specific packaging needs then we do offer paid support and you can contact sales@anaconda.com to enquire about that.

**Comment by nick-youngblut** [comment]
I've contacted sales@anaconda.com about getting paid support for our research group, but setting up payments for some sort of contract through my institute will probably take weeks if not months. A deal-breaker for using Rstudio-Server instead of Jupyter Notebooks is the lack of concurrent sessions (multiple "notebooks" running in parallel). This is only offered in the $10k/year version of RStudio-Server, and setting up a payment will take weeks or months. 

So, I'm back to using conda + Jupyter + R, and I have another (potentially dumb) problem. I can create a new issue if needed. `--cran-url https://mran.microsoft.com/snapshot/2018-01-01` fixed the `An HTTP error occurred` problem, but with `batchtools`, I'm getting a dependency conflict, which I can't figure out. I've tried building `batchtools` against CRAN or MRAN, and both have dependency conflicts:

MRAN:
```
  - r-base64url[version='>=1.1']
  - r-checkmate[version='>=1.8.5']
```

CRAN:
```
  - r-checkmate[version='>=1.8.5']
  - r-progress[version='>=1.1.1']
```

`conda info r-base64url` doesn't generate anything, but running `conda info` on ` r-checkmate` and `r-progress` indicates a conflict with R-base. But the `meta.yaml` file for both packages doesn't specify a specific R-base version (at least, that I see). I've looked through the [conda-build docs](https://conda.io/docs/user-guide/tasks/build-packages/index.html) and the help for `conda-build` and `conda skeleton cran`, but I haven't been able to figure out why there's a dependency conflict and how to fix it. 

I'd appreciate any help that you can provide.
